propagation_ot.go

```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
 	if !ok {
 		return nil, opentracing.ErrInvalidCarrier
 	}
-	// Handle the trace, span ids, and sampled status.
-	var traceID, propagatedSpanID int64
-	var sampledByte byte
 
-	if err := binary.Read(carrier, binary.BigEndian, &traceID); err != nil {
-		if err == io.EOF {
-			return nil, opentracing.ErrTraceNotFound
-		}
-		return nil, opentracing.ErrTraceCorrupted
-	}
-	if err := binary.Read(carrier, binary.BigEndian, &propagatedSpanID); err != nil {
+	// Read the length of marshalled binary. io.ReadAll isn't that performant
+	// since it keeps resizing the underlying buffer as it encounters more bytes
+	// to read. By reading the length, we can allocate a fixed sized buf and read
+	// the exact amount of bytes into it.
+	var len uint32
+	if err := binary.Read(carrier, binary.BigEndian, &len); err != nil {
```

Comment:
@bg451 this is complex enough that I think I'm sold on `fixed32`... That said, I have a few thoughts about this...

Really we should be talking about the decode – not encode – function here:

https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/decode.go#L57

```
// DecodeVarint reads a varint-encoded integer from the slice.
// It returns the integer and the number of bytes consumed, or
// zero if there is not enough.
// This is the format for the
// int32, int64, uint32, uint64, bool, and enum
// protocol buffer types.
func DecodeVarint(buf []byte) (x uint64, n int) { ... }
```

So, per the "if there's not enough" bit, you could read progressively longer byte slices until you're able to decode the size. Or there's a more streaming-ish impl. It basically looks a lot like reading UTF-8 chars, etc... easy to do in a streaming fashion.

But yeah: I'm fine with punting and doing fixed32 for the sake of simplicity.